lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp

```diff
@@ -204,6 +206,11 @@ void IMDeadCodeElimPass::runOnOperation() {
   mlir::parallelForEach(circuit.getContext(),
                         circuit.getBody()->getOps<FModuleOp>(),
                         [&](auto op) { rewriteModuleBody(op); });
+
+  // Erase empty modules. It is necessary to delete in the post order.
```

Comment:
Is it?  IR can be invalid during a pass.

Reply:
Post-order is necessary to get optimal results. For example, we have to visit modules in the order `B -> A -> Top` in the example below to delete instances optimally. 

```scala
circuit Top:
  module B:
    // empty
  module A:
    inst b of B
  module Top:
    inst a of A
```

Reply:
but we've computed the liveness in the parallel region haven't we?  This is just deleting things marked to-delete, not computing their liveness?

Reply:
> but we've computed the liveness in the parallel region haven't we? This is just deleting things marked to-delete, not computing their liveness?

No, previously liveness of modules are not calculated. I changed to propagate liveness of modules but TBH I feel it makes implementation more complicated.